### PR TITLE
*: move column, constant, like typeinfer to expression rewrite

### PR DESCRIPTION
**Pull request by XuHuaiyu**
move typeinferer of column, constant and builtin func like to expression_rewriter.go

### CODE REVIEW

File: expression/builtin_like.go [hunk 1]
```diff
@@ -37,48 +37,50 @@ type likeFunctionClass struct {
...
-	}
-
-	valStr, err := args[0].ToString()
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinLikeSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	valStr, isNull, err := b.args[0].EvalString(row, sc)
+	if isNull || err != nil {
+		return 0, true, errors.Trace(err)
```
Comment (hanfei1991) [inline expression/builtin_like.go]: should return `0, isNull, errors.Trace(err)`. isNull may not be true

File: expression/builtin_like.go [hunk 2]
```diff
@@ -37,48 +37,50 @@ type likeFunctionClass struct {
...
 	// TODO: We don't need to compile pattern if it has been compiled or it is static.
-	if args[1].IsNull() {
-		return
-	}
-	patternStr, err := args[1].ToString()
-	if err != nil {
-		return d, errors.Trace(err)
+	patternStr, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return 0, true, errors.Trace(err)
```
Comment (hanfei1991) [inline expression/builtin_like.go]: ditto

File: expression/builtin_like.go [hunk 3]
```diff
@@ -37,48 +37,50 @@ type likeFunctionClass struct {
...
 	}
 	var escape byte = '\\'
 	// If this function is called by mock tikv, the args len will be 2 and the escape will be `\\`.
 	// TODO: Remove this after remove old evaluator logic.
-	if len(args) >= 3 {
-		escape = byte(args[2].GetInt64())
+	if len(b.args) >= 3 {
+		val, isNull, err := b.args[2].EvalInt(row, sc)
+		if isNull || err != nil {
+			return 0, true, errors.Trace(err)
```
Comment (hanfei1991) [inline expression/builtin_like.go]: ditto

File: expression/builtin_like.go [hunk 4]
```diff
@@ -37,48 +37,50 @@ type likeFunctionClass struct {
...
+	sig := &builtinLikeSig{baseIntBuiltinFunc{bf}}
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
 type builtinLikeSig struct {
-	baseBuiltinFunc
+	baseIntBuiltinFunc
 }
 
 // eval evals a builtinLikeSig.
```
Comment (zz-jason) [inline expression/builtin_like.go]: update comment, change `eval` to `evalInt`

File: expression/builtin_like.go [hunk 5]
```diff
@@ -37,48 +37,50 @@ type likeFunctionClass struct {
...
-	patternStr, err := args[1].ToString()
-	if err != nil {
-		return d, errors.Trace(err)
+	patternStr, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
 	}
 	var escape byte = '\\'
 	// If this function is called by mock tikv, the args len will be 2 and the escape will be `\\`.
 	// TODO: Remove this after remove old evaluator logic.
```
Comment (zz-jason) [inline expression/builtin_like.go]: can the TODO be done ?

File: plan/typeinfer_test.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
+package plan_test
```
Comment (lamxTyler) [inline plan/typeinfer_test.go]: Add license?

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991 @zz-jason

**Comment by XuHuaiyu** [comment]
PTAL @lamxTyler

File: expression/builtin_like.go (hunk 5 above)
Comment (XuHuaiyu) [inline expression/builtin_like.go]: Noop, still needed.

**Pull request merged**